src/node_crypto.cc

```diff
@@ -5029,197 +5702,59 @@ class GenerateKeyPairJob : public CryptoJob {
-                     unsigned int n_opts,
+                     unsigned int offset,
                      std::unique_ptr<KeyPairGenerationConfig> config) {
   Environment* env = Environment::GetCurrent(args);
-  PublicKeyEncodingConfig public_key_encoding;
-  PrivateKeyEncodingConfig private_key_encoding;
-
-  // Public key encoding: type (int) + pem (bool)
-  CHECK(args[n_opts]->IsInt32());
-  public_key_encoding.type_ = static_cast<PKEncodingType>(
-      args[n_opts].As<Int32>()->Value());
-  CHECK(args[n_opts + 1]->IsInt32());
-  public_key_encoding.format_ = static_cast<PKFormatType>(
-      args[n_opts + 1].As<Int32>()->Value());
-
-  // Private key encoding: type (int) + pem (bool) + cipher (optional, string) +
-  //                       passphrase (optional, string)
-  CHECK(args[n_opts + 2]->IsInt32());
-  private_key_encoding.type_ = static_cast<PKEncodingType>(
-      args[n_opts + 2].As<Int32>()->Value());
-  CHECK(args[n_opts + 1]->IsInt32());
-  private_key_encoding.format_ = static_cast<PKFormatType>(
-      args[n_opts + 3].As<Int32>()->Value());
-  if (args[n_opts + 4]->IsString()) {
-    String::Utf8Value cipher_name(env->isolate(),
-                                  args[n_opts + 4].As<String>());
-    private_key_encoding.cipher_ = EVP_get_cipherbyname(*cipher_name);
-    if (private_key_encoding.cipher_ == nullptr)
-      return env->ThrowError("Unknown cipher");
-
-    // We need to take ownership of the string and want to avoid creating an
-    // unnecessary copy in memory, that's why we are not using String::Utf8Value
-    // here.
-    CHECK(args[n_opts + 5]->IsString());
-    Local<String> passphrase = args[n_opts + 5].As<String>();
-    int len = passphrase->Utf8Length(env->isolate());
-    private_key_encoding.passphrase_length_ = len;
-    void* mem = OPENSSL_malloc(private_key_encoding.passphrase_length_ + 1);
-    CHECK_NOT_NULL(mem);
-    private_key_encoding.passphrase_.reset(static_cast<char*>(mem),
-        [len](char* p) {
-          OPENSSL_clear_free(p, len);
-        });
-    passphrase->WriteUtf8(env->isolate(),
-                          private_key_encoding.passphrase_.get());
-  } else {
-    CHECK(args[n_opts + 5]->IsNullOrUndefined());
-    private_key_encoding.cipher_ = nullptr;
-    private_key_encoding.passphrase_length_ = 0;
-  }
+
+  PublicKeyEncodingConfig public_key_encoding =
+      GetPublicKeyEncodingFromJS(args, &offset, kKeyContextGenerate);
+  NonCopyableMaybe<PrivateKeyEncodingConfig> private_key_encoding =
+      GetPrivateKeyEncodingFromJS(args, &offset, kKeyContextGenerate);
+
+  if (private_key_encoding.IsEmpty())
+    return;
 
   std::unique_ptr<GenerateKeyPairJob> job(
       new GenerateKeyPairJob(env, std::move(config), public_key_encoding,
-                             private_key_encoding));
-  if (args[n_opts + 6]->IsObject())
-    return GenerateKeyPairJob::Run(std::move(job), args[n_opts + 6]);
+                             std::move(private_key_encoding.Release())));
```

Comment:
IIUC `private_key_encoding.Release()` is already an rvalue, so you don't need to `mode` it.